arangod/V8Server/v8-voccursor.cpp

```diff
@@ -157,6 +164,386 @@ static void JS_JsonCursor(v8::FunctionCallbackInfo<v8::Value> const& args) {
+
+  static void hasNext(v8::FunctionCallbackInfo<v8::Value> const& args) {
+    TRI_V8_TRY_CATCH_BEGIN(isolate);
+    v8::HandleScope scope(isolate);
+
+    V8Cursor* self = V8Cursor::unwrap(args.Holder());
+    if (self == nullptr) {
+      TRI_V8_RETURN_UNDEFINED();
+    }
+    
+    // we always need to fetch
+    if (!self->maybeFetchBatch(isolate)) { // sets exception
+      return;
+    }
+    
+    if (self->_dataIterator != nullptr) {
+      TRI_V8_RETURN_TRUE();
+    }
+    TRI_V8_RETURN_FALSE();
+    TRI_V8_TRY_CATCH_END
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////
+  /// @brief ArangoQueryStreamCursor.prototype.next = ...
+  ////////////////////////////////////////////////////////////////////////////////
+
+  static void next(v8::FunctionCallbackInfo<v8::Value> const& args) {
+    TRI_V8_TRY_CATCH_BEGIN(isolate);
+    v8::HandleScope scope(isolate);
+    
+    V8Cursor* self = V8Cursor::unwrap(args.Holder());
+    if (self == nullptr) {
+      TRI_V8_RETURN_UNDEFINED();
+    }
+    
+    // we always need to fetch
+    if (!self->maybeFetchBatch(isolate)) { // sets exception
+      return;
+    }
+    
+    if (self->_dataIterator) { // got a current batch
+      TRI_ASSERT(self->_dataIterator->valid());
+      
+      VPackSlice s = self->_dataIterator->value();
+      v8::Local<v8::Value> val = TRI_VPackToV8(isolate, s, &self->_options);
+      
+      ++(*self->_dataIterator);
+      // reset so that the next one can fetch again
+      if (!self->_dataIterator->valid()) {
+        self->_dataIterator.reset();
+      }
+      TRI_V8_RETURN(val);
+    }
+    
+    TRI_V8_RETURN_UNDEFINED();
+    TRI_V8_TRY_CATCH_END
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////
+  /// @brief ArangoQueryStreamCursor.prototype.count = ...
+  ////////////////////////////////////////////////////////////////////////////////
+
+  static void count(v8::FunctionCallbackInfo<v8::Value> const& args) {
+    TRI_V8_TRY_CATCH_BEGIN(isolate);
+    v8::HandleScope scope(isolate);
+    
+    V8Cursor* self = V8Cursor::unwrap(args.Holder());
+    if (self == nullptr) {
+      TRI_V8_RETURN(v8::Undefined(isolate));
+    }
+    
+    TRI_V8_RETURN_UNDEFINED(); // always undefined
+    TRI_V8_TRY_CATCH_END
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////
+  /// @brief explicitly discard cursor, mostly relevant for testing
+  ////////////////////////////////////////////////////////////////////////////////
+  
+  static void dispose(v8::FunctionCallbackInfo<v8::Value> const& args) {
+    TRI_V8_TRY_CATCH_BEGIN(isolate);
+    V8Cursor* self = V8Cursor::unwrap(args.Holder());
+    if (self != nullptr) {
+      TRI_GET_GLOBALS();
+      CursorRepository* cursors = v8g->_vocbase->cursorRepository();
+      cursors->remove(self->_cursorId, Cursor::CURSOR_VPACK);
+    }
```

Comment:
Should this function additionally reset some internal state, such as `_dataIterator`, `_dataSlice`, `_extraSlice`, `_hasMore`? I think this would make sense.